### DISCUSSION TITLE
Announcement: Per-Project Security Teams

### DISCUSSION THREAD

**Discussion by mbarbero** [eclipse-csi/.github#4] [Announcements]
In response to requests from various projects and after discussions between the Eclipse Foundation Security Team and the Architecture Council, we are pleased to announce the creation of Project Security Teams.

Project Security Teams allow projects to explicitly designate individuals responsible for handling vulnerability reports. Project Leads can define membership in [PMI](https://projects.eclipse.org/).

This initiative represents a significant step forward in enhancing visibility and communication around security issues within the Eclipse Foundation community.

If all Committers in your project are involved in addressing security issues, nothing will change for you. All Committers will automatically be considered part of your Project Security Team, and no further action is required on your part.

However, if your project has a more complex structure with only a limited number of individuals managing vulnerability reports, Project Leads can establish a dedicated Project Security Team. 

Additionally, please be aware that in early September, all members of Project Security Teams for projects hosted on GitHub (in an organization other than https://github.com/eclipse) will be granted the [Security Manager role](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/managing-security-managers-in-your-organization). If your project does not make any changes in PMI by then, and in line with the default setting where the Project Security Team equals all Committers, all Committers will be granted the Security Manager role in their respective organization on GitHub.

The Foundation's policy of openness remains unchanged: all security issues will continue to be eventually disclosed, and the Eclipse Foundation Security Team will ensure that this practice continues.

For detailed information on the introduction of Project Security Teams, including the specific permissions granted to members, please refer to the [updated Handbook](https://www.eclipse.org/projects/handbook/#projects-security-team).

_(announcement also posted to the eclipse.org-committers mailing list https://www.eclipse.org/lists/eclipse.org-committers/msg01448.html)_

**Comment by scottslewis** [comment]
>However, if your project has a more complex structure with only a limited number of individuals managing vulnerability reports, >Project Leads can establish a dedicated Project Security Team.

Questions:  

How can project leads without resources establish such a Security Team?   
Who will be on the Project Security Team?  
What will be done for security issues that are from the use of third-party libraries?

**Reply by mbarbero** [reply]
> How can project leads without resources establish such a Security Team?
> Who will be on the Project Security Team?

For projects where resources are limited, the default approach is that all committers automatically become part of the Project Security Team. While we don't mandate the creation of a dedicated Security Team by Project Leads, we do want to offer this option for those projects that have the capacity and desire to form one.

> What will be done for security issues that are from the use of third-party libraries?

I'm having difficulty seeing the connection between this question and the topic of forming a Project Security Team. Could you please clarify what you mean by "what will be done"? Are you asking about how these security issues will be handled by the Project Security Team, or something else?

**Reply by scottslewis** [reply]
> > How can project leads without resources establish such a Security Team?
> > Who will be on the Project Security Team?
> 
> For projects where resources are limited, the default approach is that all committers automatically become part of the Project Security Team. While we don't mandate the creation of a dedicated Security Team by Project Leads, we do want to offer this option for those projects that have the capacity and desire to form one.

And for those that have the desire without the capacity?  My project is ECF, which provides the install/update transport for p2.  I/we do not have enough resources now (my time) to maintain/update the filetransfer component that we contribute to the IDE.

Of course, the instal/update/transport/network communication for install/update is a critical locus of security issues for the IDE.

It's not feasible that by making me a SecurityManager, that I will be able to deal with the IDE's transport security issues.

> 
> > What will be done for security issues that are from the use of third-party libraries?
> 
> I'm having difficulty seeing the connection between this question and the topic of forming a Project Security Team. Could you please clarify what you mean by "what will be done"? Are you asking about how these security issues will be handled by the Project Security Team, or something else?

To get specific:   ECF contributes a filetransfer provider to the IDE to support p2 install/update.   Currently, ECF exposes two https-based providers:  one based upon Apache Httpclient 5, one based upon the Java 9+ JRE.   The JRE-one is newer and intended to replace the Apache 5-based provider.  Both of these providers, however, are regularly found to have security vulnerabilities, which at the very least requires a version bump, testing for regression, build/deploy cycle for ECF project.   This itself has become unsustainable for me personally and so I'm contemplating not doing it any longer.

So, for ECF/me that's the relationship between a 'security team' (just me/existing committers it sounds like) and security maintenance for the IDE file transfer transport.

**Reply by mbarbero** [reply]
Thank you, Scott, for your detailed response. I sympathize with your situation. Unfortunately, the primary goal of allowing projects to establish a Project Security Team is not to help projects cope with limited resources, although that might be a beneficial side effect. Project Leads will now have the option to invite non-committers they trust and who possess the necessary skills to assist with handling vulnerability reports. Additionally, by granting Project Security Team members the GitHub Security Manager role, they will gain access to tools that are closer to the code (i.e., on GitHub), ultimately reducing the time spent dealing with vulnerabilities.

That said, this new feature does not impose any additional burden on projects that prefer to continue operating as they were before. Projects still need to comply with the (soon to be revised to include the Project Security Team concept) [Eclipse Foundation Security Policy](https://www.eclipse.org/security/policy/), and the [Eclipse Foundation Security Team](https://www.eclipse.org/security/team/) will continue to assist with handling reports, communicating with security researchers, and publishing advisories.

Finally, I would recommend reaching out to the [Eclipse IDE Working Group](https://eclipseide.org/working-group/) to express your concerns about ECF. They might be able to secure funding for some [initiatives](https://eclipseide.org/working-group/development-funding-initiative/) related to ECF.

**Reply by scottslewis** [reply]
> Thank you, Scott, for your detailed response. I sympathize with your situation. Unfortunately, the primary goal of allowing projects to establish a Project Security Team is not to help projects cope with limited resources, although that might be a beneficial side effect. 

Yes it could be a positive 'side effect'.  I think it would be more project-and-community valuable to have it not be a 'side effect' at all...but rather to provide actual resources to make Eclipse/and other project security better...rather than what I consider more 'unfunded/unresourced mandates' on the simrel projects.

Actually, as @jerboaa suggests, for many projects making the default to be all committers may not make sense at all.  Rather, it would make more sense  to have security expertise...read existing and/or new project committers...who are certainly the only ones capable of actually dealing with any security problems anyway, resourced to assist when/if a security vulnerability is identified.  But no...as usual the EF is going to quietly give the existing project committers more responsibility/work/requirements, with only the corporate-controlled projects with enough project-level technical resources (people) to do something about it.

> 
> That said, this new feature does not impose any additional burden on projects that prefer to continue operating as they were before. Projects still need to comply with the (soon to be revised to include the Project Security Team concept) [Eclipse Foundation Security Policy](https://www.eclipse.org/security/policy/), and the [Eclipse Foundation Security Team](https://www.eclipse.org/security/team/) will continue to assist with handling reports, communicating with security researchers, and publishing advisories.

In ECF's case (comm transport for p2/Eclipse IDE install/update), none of this is helpful....here's why:  If there is a security issue with...e.g...http/https-level communication for install update in Eclipse, I am quite sure...based upon years of history...that the IDE project will expect and require that ECF deal with it (e.g. by patching old versions of Apache httpclient5 or changing the way that auth is done, dealing with ssl and/or socket-level issues...whatever it is).  It matters very little to me that there's someone to handle the reporting, communicate with security researchers, publishing advisories, etc...it will likely be my responsibility to apply whatever fix is need (whether in ECF code or not) to the ECF filetransfer codebase.  This as least partially because I'm probably the only remaining committer that has enough familiarity with this code and this has been regularly happening for or...say...10 years now.

> 
> Finally, I would recommend reaching out to the [Eclipse IDE Working Group](https://eclipseide.org/working-group/) to express your concerns about ECF. They might be able to secure funding for some [initiatives](https://eclipseide.org/working-group/development-funding-initiative/) related to ECF.

Oh...I have been engaged in a multi-year dialog with the IDE working group, committer board reps, and others about this.  FWIW, there is actually a high-importance issue in the IDE working group queue:  https://gitlab.eclipse.org/eclipse-wg/ide-wg/ide-wg-dev-funded-efforts/ide-wg-dev-funded-program-planning-council-top-issues/-/boards/1208.  

Sadly, this issue is simply for *completing* the work on a provider contributed by another committer (jre-based provider), it won't provide any resources for present or future needs for maintenance (whether bugs or security issues).  I've had many discussions about providing such support to ECF for IDE maintenance/security (and other resource-starved projects at EF...of which there are many others than just ECF I expect), with the IDE people directly, with the IDE working group people, the committer Board reps and others.    It's resulted in nothing on the real problem for ECF and many other projects at EF: paid committer time for maintenance.

Just for everyone's information: I'm not able nor willing to continue in current mode wrt the ECF filetransfer contribution to the IDE.  I don't want that to be a secret when/if I become unavailable.

**Reply by mbarbero** [reply]
> as usual the EF is going to quietly give the existing project committers more responsibility/work/requirements, with only the corporate-controlled projects with enough project-level technical resources (people) to do something about it.

I'm sorry to hear that you feel this way. We've given this situation careful consideration, as mentioned in the announcement, including consulting with the Architecture Council. Our goal is not to burden existing project committers with additional responsibilities, but rather to empower them by providing access to resources they previously couldn't access—such as through the GitHub Security Manager role. The reason committers didn't have this access before was due to the lack of flexibility in selecting a restricted, need-to-know group within projects. The introduction of the Project Security Team concept makes this feasible, allowing projects to assign security-related tasks to a subset of committers with the right expertise to handle vulnerability reports.

I understand that the resource constraints of the ECF project are challenging. However, this discussion might not be the right place to address those concerns. The Eclipse Foundation does offer avenues for project funding, such as through the [Members Funded Initiative](https://www.eclipse.org/org/workinggroups/mfi_program.php) within Industry Working Groups. I encourage you to continue raising your concerns through the IDE Working Group.

**Reply by scottslewis** [reply]
>Our goal is not to burden existing project committers with additional responsibilities, but rather to empower them by providing access to resources they previously couldn't access—such as through the GitHub Security Manager role. 

And I'm telling you/EF:  no matter how many 'authorities' you consult, you ware not 'empowering' anyone.   By denying them the resources that are actually needed (actual people who know how to deal with the issues that inevitably arise), you are making it worse by throwing on more 'management'.

> The reason committers didn't have this access before was due to the lack of flexibility in selecting a restricted, need-to-know group within projects. The introduction of the Project Security Team concept makes this feasible, allowing projects to assign security-related tasks to a subset of committers with the right expertise to handle vulnerability reports.

Right.   That works really well when there are people (not 'projects') with that expertise available to do the actual work (instead of...or at least in preference too...the 'management' of that work).  And for those projects (not just ECF) that have no 'subset of committers with the right expertise'?  (see @fedejeanne comment today)?

Here's an idea:  train a AI model to fix security issues and do bug fixing/maintenance of the platform.  Maybe the corps will support that.

> 
> I understand that the resource constraints of the ECF project are challenging. 
>However, this discussion might not be the right place to address those concerns. 

As I said, I've been personally making these points to the working group and others for literally years now to deaf ears.  I'm finished with that.  It's now the IDE people, EF people, and the committer people's problem, not mine.

**Reply by scottslewis** [reply]
Relevant to our discussion about security and open source maintenance:

https://www.theregister.com/2024/09/18/open_source_maintainers_underpaid/

**Reply by mbarbero** [reply]
I’ve read Tidelift's report thoroughly. Please take a look at the 'Fellowship for Maintainers' program at Sovereign Tech Fund (https://www.sovereigntechfund.de/programs/fellowship). It might be an interesting opportunity for you.

**Reply by scottslewis** [reply]
> I’ve read Tidelift's report thoroughly. Please take a look at the 'Fellowship for Maintainers' program at Sovereign Tech Fund (https://www.sovereigntechfund.de/programs/fellowship). It might be an interesting opportunity for you.

Thank you @mbarbero for this reference.   I have applied for a fellowship at STF for ECF including the ECF filetransfer/p2/Eclipse IDE maintenance, and including ECF's transport independent and multi-provider impl of OSGi Remote Services specification.   If this support comes through, then it will allow me to maintain ECF's contributions to the Eclipse IDE (filetransfer/transport-level security), other projects and consumers that use other parts of ECF (e.g. remote services), and the Project Security Teams effort.

**Reply by mbarbero** [reply]
Thank you for the update. I'm glad this helped and sincerely hope it will work for you!

**Comment by jerboaa** [comment]
> If all Committers in your project are involved in addressing security issues, nothing will change for you. All Committers will automatically be considered part of your Project Security Team, and no further action is required on your part.

Do I understand this correctly that the default Project Security Team will be all committers? Wouldn't that be a wrong default, going against the best practise of strictly need-to-know on security issues? Would it make sense to make the default only be the project leads - who would then be able to loop in required experts and only those? Thoughts?

**Reply by mbarbero** [reply]
> Do I understand this correctly that the default Project Security Team will be all committers?

Yes, that is correct.

> Wouldn't that be a wrong default, going against the best practise of strictly need-to-know on security issues? Would it make sense to make the default only be the project leads - who would then be able to loop in required experts and only those? Thoughts?

We also recommend limiting the Project Security Team to a small, need-to-know group of individuals. However, this decision should be made at the project level to align with the Eclipse Foundation Development Process, which emphasizes the following:

> _All Project Committers have equal rights and responsibilities within the Project. Partitioning of responsibility within a Project is managed using social convention._  
— [Eclipse Foundation Development Process, Section 4.1](https://www.eclipse.org/projects/dev_process/#4_1_Committers)

and

> _Project Leads are responsible for ensuring that their Project’s Committers adhere to the Eclipse Foundation Development Process and that the Project is engaging in activities that foster vibrant communities of users, adopters, and contributors._  
— [Eclipse Foundation Development Process, Section 4.6.2](https://www.eclipse.org/projects/dev_process/#4_6_2_PL)

Given this framework, restricting access to certain information, such as vulnerability reports, is not something that can be universally enforced across all projects. Instead, it is a decision to be made by each project.

Project leads do have the option to implement what you’ve suggested. After consulting with the committers, they can go to their project's PMI page, uncheck the _committers_ box, and check the _project leads_ box to define the membership of the Project Security Team accordingly:

![image](https://github.com/user-attachments/assets/a5c15871-bef7-40c1-896a-b89c94ffbedd)

I hope this clarifies our reasoning for making all committers the default members of the Project Security Teams.

**Reply by jerboaa** [reply]
OK. Thanks for this clarification. Much appreciated. I second the recommendation for leads to manually change this - unfortunate - default.

**Comment by fedejeanne** [comment]
Hello @mbarbero,
thank you very much for your (and your teams') continuous work on this important topic! 

Are there any trainings you recommend, or someone we (committers) can reach out to to ask and learn how to solve the issues? I have the feeling that most of the issues would be solvable if the teams started building expertise, but the whole topic of security is sometimes daunting and sometimes simply ignored because "who has the time?", right?

Don't get me wrong: I consider it to be very important, but I never find the time to start going about it and I feel I'm not the only one on that boat. So maybe some guidance (links to articles, trainings, learning paths in some study platform, etc) would help. Is there any recommendation from the security team about this? Or from the EF itself?

**Reply by mbarbero** [reply]
Thanks, @fedejeanne, for your kind words. This is truly a team effort, and I'll be sure to share the kudos with everyone involved.

For any security-related questions, you can always reach out to the Eclipse Foundation Security Team. If the topic shouldn't be discussed publicly, you can send an email to security@eclipse-foundation.org. Otherwise, feel free to use [this forum](https://github.com/orgs/eclipse-csi/discussions/) to ask your questions.

Regarding training, we've started compiling information in what we call the Eclipse Foundation Security Handbook: https://eclipse-csi.github.io/security-handbook/. This is still a work in progress, but the content will grow over time.

We will also begin offering committer training in Q4 2024 through a series of webinars. @mrybczyn can provide more details about this.

Do you think this combination of written materials and live training (with recordings available) would meet your needs?

**Reply by fedejeanne** [reply]
Thank you for the pointers! I'll be definitely taking a look at what you mentioned and I'll be looking forward to the webinars from Marta :-)

> Do you think this combination of written materials and live training (with recordings available) would meet your needs?

I think it would 👍

**Reply by mrybczyn** [reply]
@fedejeanne I will jump on the opportunity :) Do you have ideas of subjects that you would like to see covered first?

**Reply by fedejeanne** [reply]
> @fedejeanne I will jump on the opportunity :) Do you have ideas of subjects that you would like to see covered first?

Well in my particular case, information on how to assess the **real** risk of a finding would be useful. I understand the theory of the findings just fine (they are thoroughly described in the advisories) but I struggle to answer questions such as:
* Is this really reproducible in our environment?
* If it is, which level of access should an attacker already posses in order to exploit this vulnerability? (which begs the following question:)
* If someone needs access to my computer/environment just to exploit this vulnerability, isn't all the damage already made?

Take SSRF for example: 
* CodeQL: https://codeql.github.com/codeql-query-help/javascript/js-request-forgery/
* OWASP: https://owasp.org/www-community/attacks/Server_Side_Request_Forgery

It is clear to me that if the URL is manipulated then bad things may happen. The part I struggle with is: if the URL is in `localhost` and the service "simply" displays the documentation in an embedded browser then:
1. Is it worth protecting a very simple server that merely serves HTML documentation?
2. Doesn't the attacker have to be in the system before even being able to exploit such a vulnerability? 

In my very reduced understanding I would answer like this:
1. It's not worth it, worst case scenario the documentation is wrong and I have to search somewhere else
2. Yes. All hope is lost already.

But an expert might answer very differently. Which is why I usually leave those findings unattended out of fear of leaving some big security hole unattended :-\ 

I hope I could answer your question @mrybczyn :-)

**Comment by mbarbero** [comment]
Just a heads up: Project Security Team members have now been granted the [Security Manager](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/managing-security-managers-in-your-organization) role within their respective GitHub organizations.

**Comment by scottslewis** [comment]
For everyone's information...wrt open source security trends:

https://www.theregister.com/AMP/2024/12/10/ai_slop_bug_reports/

**Comment by scottslewis** [comment]
This comment is intended to provide awareness among EF project leads, committers, and contributors

In fall 2024, the EU passed the Cyber Resilience Act:  https://en.wikipedia.org/wiki/Cyber_Resilience_Act

The Eclipse Foundation has created an Open Regulatory Working Group for orgs and committers collaborate on compliance:  https://orcwg.org/.  See also Mike's Blog and posts:  https://blogs.eclipse.org/post/mike-milinkovich/securing-future-open-source-launching-open-regulatory-compliance-working.  

The CRA has a new concept of an 'open source steward' that is intended to shield free and open source projects and communities from some of the liability

 The OR working group has created a mailing list here:

https://accounts.eclipse.org/mailing-list/open-regulatory-compliance 

...and there has been some recent discussion on this list of how the open source steward role is defined by the CRA, and what stewards, manufacturers, orgs, and others will be responsible for in terms of compliance.   It's very early in figuring these things out...for the EF umbrella of OSS communities, but a good time for open source community input.
